🎨 Added a configuration option for the paging parameter name

SebastianSchroeder opened the pull request:
I created this PR as I want to use Ghost for my German website and the option to configure the name of the paging parameter was missing. Currently I have to live with www.geist.de/page/1 when I would rather use www.geist.de/seite/1 as that matches the language of my site. 

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds a configurable `pagination.pageParameter` and updates URL generation, routing, and template logic to use it.
> 
> - **Config**:
>   - Adds `pagination.pageParameter` (default: `"page"`) in `core/shared/config/defaults.json`.
> - **Frontend**:
>   - **URL Generation**: `core/frontend/meta/paginated-url.js` builds paginated URLs using configured page parameter and regex.
>   - **Routing Middleware**: `services/routing/middleware/page-param.js` validates/redirects using custom parameter and sets `req.params[<param>]`.
>   - **Controllers**: `services/routing/controllers/{channel,collection}.js` read `req.params[<param>]` to compute `page` and limits.
>   - **Rendering**:
>     - `services/rendering/context.js` detects `paged` context via custom parameter.
>     - `services/rendering/templates.js` passes `req.params[<param>]` to template resolution.
> - **Tests**:
>   - Extend unit tests to stub `config.get('pagination:pageParameter')` (e.g., `"seite"`) and verify URL/build/render behavior across `paginated-url`, `channel`, and `collection`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 188037c0b47d88914d4628c63aab0cb1e8b7054a. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

SebastianSchroeder committed:
🎨 Added a configuration option for the paging parameter name

Enables websites to be configured with a custom paging parameter name. This is especially helpful for internationalized installations where it should be a localized parameter (e.g., seite for German) instead of the standard English page parameter.

SebastianSchroeder commented:
@cathysarisky I didn't tag this with the i18n emoji but i18n is actually the reason why I opened this PR. What do you say?

cathysarisky commented:
> @cathysarisky I didn't tag this with the i18n emoji but i18n is actually the reason why I opened this PR. What do you say?

I think the PR is a great idea, but I'm not going to tag it as i18n, because if I do, there's a good chance that stalebot will close it before someone has time to review it.

cathysarisky commented:
I'm not going to be the one to review & merge this PR, @SebastianSchroeder (a bit out of scope for me), but I did want to note that (if I've understood it correctly) your approach is only going to allow an alternate word for 'page' for sites that are self-hosted or otherwise have access to their config.  It will not address the problem for folks on Ghost Pro, since they don't have access to their config files.  That doesn't make this a bad PR, but I do wonder if there's a way to accomplish the same thing that'll work in more hosting environments.

SebastianSchroeder commented:
I don't think it would be a problem to make this available in the admin ui as a configuration option. I would think though that that comes with additional requirements regarding checks against special url characters like / or ?. And a decision would need to be made where in the admin ui such an option would be placed. Maybe you can talk about that in the ghost team and let me know so I can implement the necessary changes.

9larsons commented on ghost/core/core/frontend/services/routing/middleware/page-param.js:
```diff
@@ -1,22 +1,25 @@
+    const pageParam = config.get('pagination:pageParameter');
     // routeKeywords.page: 'page'
-    const pageRegex = new RegExp('/page/(.*)?/');
+    const pageRegex = new RegExp('/' + pageParam + '/(.*)?/');
```
Fixed in the follow-up commit — collection-router.js, taxonomy-router.js, and static-routes-router.js now derive the URL segment, the :token, and the .param() registration from pagination:pageParameter at construction time, so /seite/2/ resolves and /seite/1/ → / no longer loops. Added a router-level integration test covering a custom value, the default, the redirect, and invalid-value fallback. Resolving.

9larsons commented:
Sorry for the noise here, I thought I had a solution that would work, but I'm realizing that it may be better to support the routes.yaml version instead of scoping down to config-only. I think i18n could get a little hairy.

9larsons committed:
Added routes.yaml support and router wiring for the configurable pagination segment

Sources the pagination URL segment from a top-level routes.yaml key
(default "page"), derives the route segment/token/param from it at router
construction, and validates it on upload with a fail-safe fallback at render.

9larsons commented:
Ok, I feel fairly good about the routing changes, however there are two findings:
1. Users would be advised to also add a redirect from page -> the new pagination value. This is true of many changes to routes.yaml, although it's arguably not great ux. These are 'advanced settings', after all.
2. We ought to update our official themes from `{{@site.url}}/page/2/` → `{{page_url 2}}`. This could catch other themes, and is a big reason for the redirect (+SEO or other deep links). 

Any move to supporting this via i18n/locale changes would necessitate automatic redirects. We're working on some routing changes and it's possible we'd have better support down the road. I want to run this by some others before I say this is ok to merge.

ErisDS reviewed (changes requested):
Some history here - this _was_ originally exposed as a setting, and we worked it back, with the intention of exposing it in routes.yaml - but that was ~8 years ago, it didn't land as part of the main work and demand was low.

The heart of the work to make this more configurable in this PR seems directionally correct.

What I'm ruminating on is where and how this ought to be exposed. The declared usecase is i18n, but i18n changing URLs in the system seems pretty terrible UX.

Routes.yaml doesn't seem terrible, but we are trying to sloowwwwly get rid of it (in favour of a system that's easier to reason about).  Big picture, make it simpler thinking, leads me to think about either getting rid of the word page altogether, but that's out of scope here.

One change that I would ask to revert before merging is the comment in default-routes.yaml. Instead, if we do merge this we should document the feature well, but not make a text-only change to that file.

The reason being that this would make the hash of default-routes.yaml different to routes.yaml for the majority of sites when nothing changed.
